src/event_base.h

```diff
@@ -60,65 +71,125 @@ struct AlohalyticsBaseEvent {
 };
-CEREAL_REGISTER_TYPE_WITH_NAME(AlohalyticsBaseEvent, "b");
+CEREAL_REGISTER_TYPE_WITH_NAME(AlohalyticsBaseEvent, "b")
 
 // Special event in the beginning of each block (file) sent to stats server.
 // Designed to mark all events in this data block as belonging to one user with specified id.
 struct AlohalyticsIdEvent : public AlohalyticsBaseEvent {
   std::string id;
 
+  virtual std::string ToString() const {
+    return AlohalyticsBaseEvent::ToString() + " ID: " + id;
+  }
+
   template <class Archive>
   void serialize(Archive& ar) {
     AlohalyticsBaseEvent::serialize(ar);
     ar(CEREAL_NVP(id));
   }
 };
-CEREAL_REGISTER_TYPE_WITH_NAME(AlohalyticsIdEvent, "i");
+CEREAL_REGISTER_TYPE_WITH_NAME(AlohalyticsIdEvent, "i")
 
 // Simple event with a string name (key) only.
 struct AlohalyticsKeyEvent : public AlohalyticsBaseEvent {
   std::string key;
 
+  virtual std::string ToString() const {
+    return AlohalyticsBaseEvent::ToString() + " " + key;
+  }
+
   template <class Archive>
   void serialize(Archive& ar) {
     AlohalyticsBaseEvent::serialize(ar);
     ar(CEREAL_NVP(key));
   }
 };
-CEREAL_REGISTER_TYPE_WITH_NAME(AlohalyticsKeyEvent, "k");
+CEREAL_REGISTER_TYPE_WITH_NAME(AlohalyticsKeyEvent, "k")
 
 // Simple event with a string key/value pair.
 struct AlohalyticsKeyValueEvent : public AlohalyticsKeyEvent {
   std::string value;
 
+  virtual std::string ToString() const {
+    return AlohalyticsKeyEvent::ToString() + " = " + value;
+  }
+
   template <class Archive>
   void serialize(Archive& ar) {
     AlohalyticsKeyEvent::serialize(ar);
     ar(CEREAL_NVP(value));
   }
 };
-CEREAL_REGISTER_TYPE_WITH_NAME(AlohalyticsKeyValueEvent, "v");
+CEREAL_REGISTER_TYPE_WITH_NAME(AlohalyticsKeyValueEvent, "v")
 
 // Simple event with a string key and map<string, string> value.
 struct AlohalyticsKeyPairsEvent : public AlohalyticsKeyEvent {
   std::map<std::string, std::string> pairs;
 
+  virtual std::string ToString() const {
+    std::ostringstream stream;
+    stream << AlohalyticsKeyEvent::ToString() << " [ ";
+    for (const auto &pair : pairs) {
+      stream << pair.first << '=' << pair.second << ' ';
+    }
+    stream << ']';
+    return stream.str();
+  }
+
   template <class Archive>
   void serialize(Archive& ar) {
     AlohalyticsKeyEvent::serialize(ar);
     ar(CEREAL_NVP(pairs));
   }
 };
-CEREAL_REGISTER_TYPE_WITH_NAME(AlohalyticsKeyPairsEvent, "p");
+CEREAL_REGISTER_TYPE_WITH_NAME(AlohalyticsKeyPairsEvent, "p")
+
+// Key + location.
+struct AlohalyticsKeyLocationEvent : public AlohalyticsKeyEvent {
+  alohalytics::Location location;
+
+  virtual std::string ToString() const {
+    return AlohalyticsKeyEvent::ToString() + ' ' + location.ToDebugString();
+  }
+
+  template <class Archive>
+  void serialize(Archive& ar) {
+    AlohalyticsKeyEvent::serialize(ar);
+    ar(CEREAL_NVP(location));
+  }
+};
+CEREAL_REGISTER_TYPE_WITH_NAME(AlohalyticsKeyLocationEvent, "kl")
 
+// Key=value + location.
```

Comment:
Nit: `key + value + location`?

If I'm you, I'd add a couple trait classes (`has_value<T>` and `has_location<T>`, and unified getters for those).